File: leetcode/0349-intersection-of-two-arrays/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 7 ms (85.48%)
     * MEMORY: 17.79 MB (10.8%)
    */

    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> hashset1;
        unordered_map<int, int> hashset2;
        vector<int> unique;
        for (int i : nums1) { hashset1[i] = 1; }
        for (int i : nums2) { hashset2[i] = 1; }
        for (auto i : hashset1) {
            if (hashset2.find(i.first) != hashset2.end() ) {
                unique.push_back(i.first);
            }
        }
        return unique;
    }
};

File: leetcode/0349-intersection-of-two-arrays/solution.py
class BetterSolution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        RUNTIME: 49 ms (90.56%)
        MEMORY: 13.9 MB (91.82%)
        """
        hashset = {}
        unique = []
        for i in nums1:
            hashset[i] = 1
        for j in nums2:
            if j in hashset.keys() and j not in unique:
                unique.append(j)
        return unique


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        RUNTIME: 95 ms (52.23%)
        MEMORY: 14 MB (92.6%)
        """
        hashset1 = {}
        hashset2 = {}
        for i in nums1:
            hashset1[i] = 1
        for j in nums2:
            hashset2[j] = 1
        unique = []
        for k in hashset1.keys():
            if k in hashset2.keys():
                unique.append(k)
        return unique
